processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java

```diff
@@ -553,36 +540,24 @@ else if ( propertyEntries.size() == 1 ) {
                     forgedName = ctx.getExistingMappingMethod( nestedPropertyMapping ).getName();
                 }
                 String sourceRef = forgedName + "( " + sourceParam.getName() + " )";
-                return new SourceRHS( sourceRef, getSourceType(), existingVariableNames );
-
-            }
-        }
-
-        private String getSourceElement() {
-
-            Parameter sourceParam = sourceReference.getParameter();
-            List<PropertyEntry> propertyEntries = sourceReference.getPropertyEntries();
-            if ( propertyEntries.isEmpty() ) {
-                return String.format( "parameter \"%s %s\"", sourceParam.getType(), sourceParam.getName() );
-            }
-            else if ( propertyEntries.size() == 1 ) {
-                PropertyEntry propertyEntry = propertyEntries.get( 0 );
-                return String.format( "property \"%s %s\"", propertyEntry.getType(), propertyEntry.getName() );
-            }
-            else {
-                PropertyEntry lastPropertyEntry = propertyEntries.get( propertyEntries.size() - 1 );
-                return String.format(
-                    "property \"%s %s\"",
-                    lastPropertyEntry.getType(),
-                    Strings.join( sourceReference.getElementNames(), "." )
+                SourceRHS sourceRhs = new SourceRHS( sourceParam.getName(),
+                                                     sourceRef,
+                                                     getSourcePresenceCheckerRef( sourceReference ),
+                                                     sourceType,
+                                                     existingVariableNames,
+                                                     sourceReference.toString()
                 );
+                return sourceRhs;
+
             }
         }
 
-        private String getSourcePresenceCheckerRef() {
+        private String getSourcePresenceCheckerRef( SourceReference sourceReference ) {
             String sourcePresenceChecker = null;
             if ( !sourceReference.getPropertyEntries().isEmpty() ) {
                 Parameter sourceParam = sourceReference.getParameter();
+                // TODO is first correct here?? shouldn't it be last since the remainer is checked
```

Comment:
I have a feeling its correct. It was there, but a reminder for myself. I'll tackle  this in the follow up.